File: src/lib/forensic/stats.ts
import type { Stats } from "./types";

export function statsF32(data: ArrayLike<number>): Stats {
  const n = data.length;
  if (n === 0) {
    return {
      mean: 0, median: 0, std: 0, min: 0, max: 0,
      p05: 0, p95: 0, p99: 0, energy: 0, entropy: 0,
    };
  }
  let min = Infinity;
  let max = -Infinity;
  let sum = 0;
  let energy = 0;
  for (let i = 0; i < n; i++) {
    const x = data[i]!;
    if (x < min) min = x;
    if (x > max) max = x;
    sum += x;
    energy += x * x;
  }
  const mean = sum / n;
  let varSum = 0;
  for (let i = 0; i < n; i++) {
    const d = data[i]! - mean;
    varSum += d * d;
  }
  const sorted = Float32Array.from({ length: n }, (_, i) => data[i]!);
  sorted.sort();
  const pct = (p: number) => {
    const idx = Math.round((p / 100) * (n - 1));
    return sorted[Math.min(n - 1, Math.max(0, idx))] ?? 0;
  };
  const range = Math.max(max - min, 1e-8);
  const bins = new Uint32Array(64);
  for (let i = 0; i < n; i++) {
    const b = Math.min(63, Math.floor(((data[i]! - min) / range) * 63));
    bins[b]!++;
  }
  let entropy = 0;
  for (let i = 0; i < 64; i++) {
    const c = bins[i]!;
    if (!c) continue;
    const p = c / n;
    entropy -= p * Math.log2(p);
  }
  return {
    mean,
    median: pct(50),
    std: Math.sqrt(varSum / n),
    min,
    max,
    p05: pct(5),
    p95: pct(95),
    p99: pct(99),
    energy: energy / n,
    entropy,
  };
}

export function scaleToU8(
  src: ArrayLike<number>,
  gain: number,
  percentile: number,
): { gray: Uint8Array; stats: Stats } {
  const stats = statsF32(src);
  const n = src.length;
  const gray = new Uint8Array(n);
  if (n === 0) return { gray, stats };
  const sorted = Float32Array.from({ length: n }, (_, i) => src[i]!);
  sorted.sort();
  const p = Math.min(100, Math.max(50, percentile));
  const idx = Math.round((p / 100) * (n - 1));
  const clip = Math.max(sorted[idx] ?? 1, 1e-8);
  const g = Math.max(gain, 0.01);
  for (let i = 0; i < n; i++) {
    gray[i] = Math.max(0, Math.min(255, ((src[i]! / clip) * 255 * g) | 0));
  }
  return { gray, stats };
}

export function luma(r: number, g: number, b: number): number {
  return 0.2126 * r + 0.7152 * g + 0.0722 * b;
}

export function fmt(n: number, d = 3): string {
  if (!Number.isFinite(n)) return "—";
  if (Math.abs(n) >= 1000) return n.toFixed(1);
  if (Math.abs(n) >= 10) return n.toFixed(Math.min(d, 2));
  return n.toFixed(d);
}

export function fmtBytes(n: number): string {
  if (n < 1024) return `${n} B`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} KB`;
  return `${(n / (1024 * 1024)).toFixed(2)} MB`;
}

File: public/pack/src/stats.rs
//! Descriptive statistics for residual / heatmap buffers.

#[derive(Clone, Copy, Debug, Default)]
pub struct Stats {
    pub mean: f64,
    pub median: f64,
    pub std: f64,
    pub min: f64,
    pub max: f64,
    pub p05: f64,
    pub p95: f64,
    pub p99: f64,
    pub energy: f64,
    pub entropy: f64,
}

impl Stats {
    pub fn write_f64(&self, out: &mut [f64]) {
        if out.len() < 10 {
            return;
        }
        out[0] = self.mean;
        out[1] = self.median;
        out[2] = self.std;
        out[3] = self.min;
        out[4] = self.max;
        out[5] = self.p05;
        out[6] = self.p95;
        out[7] = self.p99;
        out[8] = self.energy;
        out[9] = self.entropy;
    }
}

pub fn stats_f32(data: &[f32]) -> Stats {
    if data.is_empty() {
        return Stats::default();
    }
    let n = data.len() as f64;
    let mut min = f32::INFINITY;
    let mut max = f32::NEG_INFINITY;
    let mut sum = 0.0f64;
    let mut energy = 0.0f64;
    for &v in data {
        let x = if v.is_finite() { v } else { 0.0 };
        if x < min {
            min = x;
        }
        if x > max {
            max = x;
        }
        let d = x as f64;
        sum += d;
        energy += d * d;
    }
    let mean = sum / n;
    let mut var = 0.0f64;
    for &v in data {
        let x = if v.is_finite() { v as f64 } else { 0.0 };
        let d = x - mean;
        var += d * d;
    }
    let std = (var / n).sqrt();

    let mut sorted: Vec<f32> = data
        .iter()
        .map(|v| if v.is_finite() { *v } else { 0.0 })
        .collect();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let percentile = |p: f64| -> f64 {
        let idx = ((p / 100.0) * (sorted.len().saturating_sub(1) as f64)).round() as usize;
        sorted[idx.min(sorted.len() - 1)] as f64
    };

    // 64-bin entropy of normalized values
    let range = (max - min).max(1e-8);
    let mut bins = [0u32; 64];
    for &v in data {
        let x = if v.is_finite() { v } else { min };
        let b = (((x - min) / range) * 63.0).floor() as usize;
        bins[b.min(63)] += 1;
    }
    let mut entropy = 0.0f64;
    for c in bins {
        if c == 0 {
            continue;
        }
        let p = c as f64 / n;
        entropy -= p * p.log2();
    }

    Stats {
        mean,
        median: percentile(50.0),
        std,
        min: min as f64,
        max: max as f64,
        p05: percentile(5.0),
        p95: percentile(95.0),
        p99: percentile(99.0),
        energy: energy / n,
        entropy,
    }
}

/// Percentile clip then linear scale to 0..255.
pub fn scale_to_u8(src: &[f32], dst: &mut [u8], gain: f32, percentile: f32) -> Stats {
    let stats = stats_f32(src);
    if src.is_empty() || dst.is_empty() {
        return stats;
    }
    let p = percentile.clamp(50.0, 100.0) as f64;
    let mut sorted: Vec<f32> = src
        .iter()
        .map(|v| if v.is_finite() { *v } else { 0.0 })
        .collect();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let idx = ((p / 100.0) * (sorted.len().saturating_sub(1) as f64)).round() as usize;
    let clip = sorted[idx.min(sorted.len() - 1)].max(1e-8);
    let g = gain.max(0.01) as f32;
    let n = src.len().min(dst.len());
    for i in 0..n {
        let v = src[i].max(0.0) / clip * 255.0 * g;
        dst[i] = v.clamp(0.0, 255.0) as u8;
    }
    stats
}

pub fn luma(r: u8, g: u8, b: u8) -> f32 {
    0.2126 * r as f32 + 0.7152 * g as f32 + 0.0722 * b as f32
}
